sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/DecorrelateInnerQuery.scala

```diff
@@ -654,6 +654,25 @@ object DecorrelateInnerQuery extends PredicateHelper {
             val newProject = Project(newProjectList ++ referencesToAdd, newChild)
             (newProject, joinCond, outerReferenceMap)
 
+          case w@Window(projectList, partitionSpec, orderSpec, child) =>
+            val outerReferences = collectOuterReferences(w.expressions)
+            assert(outerReferences.isEmpty, s"Correlated column is not allowed in window " +
+              s"function: $w")
+            val newOuterReferences = parentOuterReferences ++ outerReferences
+            val (newChild, joinCond, outerReferenceMap) =
+              decorrelate(child, newOuterReferences, aggregated = true, underSetOp)
+            val newProjectList = replaceOuterReferences(projectList, outerReferenceMap)
+            val newPartitionSpec = replaceOuterReferences(partitionSpec, outerReferenceMap)
+            val newOrderSpec = replaceOuterReferences(orderSpec, outerReferenceMap)
+            val referencesToAdd = missingReferences(
+              newPartitionSpec.asInstanceOf[Seq[NamedExpression]],
```

Comment:
What about the project list and order spec?

Reply:
good point! as we don't allow correlation anywhere in the Window operator itself, it is enough to get the joinCond attributes.

Reply:
Don't we want to pass newProjectList instead of newPartitionSpec? Since those are the outputs from the window func.

Reply:
The joinCond has both the inner and outer/domain attribute references, so I think this would duplicate some refs unnecessarily. I think we can do missingReferences(newProjectList, joinCond), similar to other cases.

Reply:
done